src/main/scala/is/hail/expr/Relational.scala

```diff
@@ -1601,6 +1601,38 @@ case class TableValue(typ: TableType, globals: BroadcastRow, rvd: RVD) {
   def rdd: RDD[Row] =
     rvd.toRows
 
+  def enforceOrderingRVD: RVD = (rvd, typ.key) match {
+    case (orvd: OrderedRVD, Some(key)) =>
+      assert(orvd.typ.key.startsWith(key))
+      orvd
+    case (orvd: OrderedRVD, None) =>
+      orvd.toUnpartitionedRVD
+    case (urvd: UnpartitionedRVD, Some(key)) =>
+      val orvdType = new OrderedRVDType(key.toArray, key.toArray, typ.rowType)
+      OrderedRVD.coerce(orvdType, rvd, None, None)
+    case (urvd: UnpartitionedRVD, None) =>
+      urvd
+  }
+
+  // Note: the returned OrderedRVD may have more key fields or fewer partition
+  // key fields. The keys and nPartitionKeys arguments are regarded as
+  // constraints the returned OrderedRVD must satisfy. If rvd is already an
+  // OrderedRVD satisfying stronger constraints, we don't need to do do any
+  // work, but we also don't want to throw away the knowledge of the stronger
+  // invariants.
+  def toOrderedRVD(keys: Array[String], nPartitionKeys: Int): OrderedRVD = {
+    rvd match {
+      case orvd: OrderedRVD
+        if orvd.typ.key.startsWith(keys)
+        && orvd.typ.partitionKey.length <= nPartitionKeys =>
```

Comment:
shouldn't this be the opposite?

Reply:
Suppose the table's rvd is ordered with key `((pk1: String), k2: Int, k3: String, k3: Int)`, and we call `toOrderedRVD(Array("pk1", "pk2", "k3"), 2)`. The old range bounds would be something like `[("A", "C"], ("C", "F"], ["H", "I"]]`. That satisfies the request that all keys sharing the same pk1 and k2 will live in the same partition.